src/librustc_mir/transform/const_prop.rs

```diff
@@ -636,18 +643,40 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
                     ScalarMaybeUndef::Scalar(one),
                     ScalarMaybeUndef::Scalar(two)
                 ) => {
-                    let ty = &value.layout.ty.kind;
+                    // Value represented as a pair, but we still want to generate zsts for rest of
+                    // the fields.
+                    let layout = &value.layout;
+                    debug!("replace_with_const: value layout={:#?}", layout);
+                    debug!("replace_with_const: value field count={}", layout.fields.count());
+                    let ty = &layout.ty.kind;
                     if let ty::Tuple(substs) = ty {
+                        // Values from the pair will be popped from this vec when we see a non-zst
+                        // type in the layout
+                        let mut pair_values = vec![two, one];
+                        // Final values in the aggregate
+                        let mut aggregate_values = vec![];
+                        // Iterate fields, for zsts a zsg scalar, for others use the values from
+                        // the pair
+                        for i in 0 .. layout.fields.count() {
+                            let field_layout = layout.field(self, i).unwrap();
+                            debug!("field {} layout={:#?}", i, field_layout);
+                            let scalar = if field_layout.details.size == Size::ZERO {
```

Comment:
There's a [dedicated function](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/layout/struct.TyLayout.html#method.is_zst) for this:

```suggestion
                            let scalar = if field_layout.is_zst() {
```